qiime2/sdk/usage.py

```diff
@@ -9,59 +9,236 @@
 
-    def validate(self, inputs, outputs):
-        if not isinstance(inputs, UsageInputs):
-            raise TypeError('Must provide an instance of UsageInputs.')
-        if not isinstance(outputs, UsageOutputNames):
-            raise TypeError('Must provide an instance of UsageOutputNames.')
+    def push_record(
+        self,
+        ref: str,
+        value: typing.Union[
+            "sdk.Artifact", "sdk.Visualization", "metadata.Metadata"
+        ],
+        source: str,
+        assert_has_line_matching: typing.Callable = None,
+    ) -> ScopeRecord:
+        """
+        Update `self._records` with an entry for this record where `ref` is
+        the key and `ScopeRecord` is the value.
+
+        Parameters
+        ----------
+        ref : str
+        value : Artifact, Visualization, or Metadata
+            Data from a Usage data initialization method
+        source : str
+            The Usage method called to initialize example data
+        assert_has_line_matching : callable
+            Verify that the file at `path` contains a line matching
+            `expression` within an Artifact. See
+            `ScopeRecord.assert_has_line_matching`.
+
+        Returns
+        -------
+        record : ScopeRecord
+
+        """
+        record = ScopeRecord(ref=ref, value=value, source=source,
+                             assert_has_line_matching=assert_has_line_matching)
+        self._records[ref] = record
+        return record
 
-        _, sig = self.get_action()
+    def get_record(self, ref: str) -> ScopeRecord:
+        """
+        Get a `ScopeRecord` from the current scope.
 
-        inputs.validate(sig)
-        outputs.validate(sig)
+        Parameters
+        ----------
+        ref : str
+            The name of a `ScopeRecord`
+
+        Raises
+        ------
+        KeyError
+            If the record name isn't in the scope
+
+        Returns
+        -------
+        record : ScopeRecord
+
+        """
+        try:
+            return self.records[ref]
+        except KeyError:
+            raise KeyError('No record with ref id: "%s" in scope.' % (ref,))
 
 
 class UsageInputs:
-    def __init__(self, **kwargs):
+    """Define the inputs for a Usage example.
+    """
+
+    def __init__(
+        self,
+        **kwargs: typing.Union[
+            int,
+            bool,
+            None,
+            typing.Iterable[int],
+            typing.Callable[
+                ..., typing.Union["metadata.Metadata", "sdk.Artifact"]
+            ],
+        ]
+    ):
+        """
+        Parameters
+        ----------
+        kwargs : ExampleInputs
+            Inputs to be passed in as keyword arguments to `Usage.action`.
+        """
         self.values = kwargs
 
     def __repr__(self):
         return 'UsageInputs(**%r)' % (self.values,)
 
-    def validate(self, signature):
+    def validate(
+        self, signature: typing.Union[MethodSignature, PipelineSignature]
+    ) -> None:
```

Comment:
```suggestion
    def validate(self, signature: 'Signature') -> None:
```

Reply:
Assuming that we will eventually run something like mypy, wouldn't it be better to use actual types in these type hints, even though they are just strings atm?  If there is no actual `Signature` class or an alias, might this type hint potentially cause confusion?  Or is there a `Signature` object?  I couldn't find one though I see it (erroneously?) mentioned [here](https://github.com/qiime2/qiime2/blob/e3def226e64340d9a474f283863edb8fad57cd7f/qiime2/sdk/action.py#L82).

Reply:
Good catch, that should be `signature: 'PipelineSignature'` on line 198 above, since `PipelineSignature` is the base class for all the signature classes.